Implement hash typeclass and its interfaces

1Jajen1 opened the pull request:
WIP because its missing docs and I have some open questions.

- Some datatypes (specifically ListK/SetK etc.) override hashCode to use kotlins Any.hashCode. Should they instead point to their instance of hash() and if so should the parameter if needed be substituted with `Hash.any()`?
- Are there any datatypes I have missed? I skipped some where I was unsure (Coproduct, Sum etc) so please take a look there :)
- As for laws: Hashes only really need to be consistent (`hash(a) == hash(a)`) and hold the law if `a == b` => `hash(a) == hash(b)`. Is that enough?

@pakoito This should be ready some docs I hope :see_no_evil: 

Solves #687

pakoito commented on modules/core/arrow-test/src/main/kotlin/arrow/test/laws/HashLaws.kt:
```diff
@@ -0,0 +1,29 @@
+
+object HashLaws {
+  fun <F> laws(HF: Hash<F>, EQ: Eq<F>, cf: (Int) -> F): List<Law> =
+    listOf(
```
Because Hash inherits from Eq then this implies that:

EqLaws.laws(...) +  listOf(...

This way you can test both Eq and Hash instances in a single call, replacing the current EqLaws call for HashLaws in the existing tests.

pakoito commented:
>Some datatypes (specifically ListK/SetK etc.) override hashCode to use kotlins Any.hashCode. Should they instead point to their instance of hash() and if so should the parameter if needed be substituted with Hash.any()?

The idea is that the instance delegates to the original implementation, which in this case would be the one of the types they're wrapping: List, Set, Sequence, etc...

> Are there any datatypes I have missed?

Per Java's indication, anything with `equals` qualifies for `hashCode` too. You don't have to do all of them in one go, but do list the missing ones in a new Issue.

> Laws
The hashLaws in Scala are here: https://github.com/typelevel/cats/blob/master/kernel-laws/src/main/scala/cats/kernel/laws/HashLaws.scala. 

IMO given that we don't have universal hash calculation we should be good with the two that you added :)

See the comment I've added, to make sure that Hash instances also imply correct Eq instances.

pakoito commented on modules/docs/arrow-docs/docs/_data/menu.yml:
```diff
@@ -234,6 +234,9 @@ options:
       - title: Alternative
         url: /docs/typeclasses/alternative/
 
+      - title: Hash
```
I'd put it below `Eq` instead, to simplify that's simple, it inherits from it, and this may be somewhat common anyway :D

pakoito commented on modules/docs/arrow-docs/docs/docs/typeclasses/hash/README.md:
```diff
@@ -0,0 +1,100 @@
+val user = User("MyId", "MyName")
+
+// This is fine
+Hash.any().run { user.hash() }
```
Split this snippet into 2 ank snippets at this line, so there's an output from this line too.

pakoito commented on modules/docs/arrow-docs/docs/docs/typeclasses/hash/README.md:
````diff
@@ -0,0 +1,100 @@
+
+```kotlin:ank
+import arrow.core.*
+import arrow.typeclasses.Hash
````
Imports are per file, not per snippet, so this one is unnecessary.

pakoito commented on modules/docs/arrow-docs/docs/docs/typeclasses/hash/README.md:
````diff
@@ -0,0 +1,100 @@
+```
+
+```kotlin:ank
+import arrow.typeclasses.Hash
````
Same duped import here.

pakoito commented on modules/docs/arrow-docs/docs/docs/typeclasses/hash/README.md:
````diff
@@ -0,0 +1,100 @@
+
+```kotlin:ank:replace
+import arrow.reflect.*
+import arrow.typeclasses.Hash
````
This one may not be necessary either.

pakoito commented on modules/core/arrow-instances-data/src/main/kotlin/arrow/instances/nonemptylist.kt:
```diff
@@ -139,6 +139,17 @@ interface NonEmptyListSemigroupKInstance : SemigroupK<ForNonEmptyList> {
+  override fun EQ(): Eq<A> = HA()
+
+  override fun NonEmptyList<A>.hash(): Int = foldLeft(1) { hash, a ->
+    31 * hash + HA().run { a.hash() }
```
@raulraja any opinions about this one?

1Jajen1 commented:
I know :) You are too fast :D Was about to push the fix for that. I also added instances for Tuple, Coproduct (derived from either) and sum. And added the tests for them as well. If they are not correct or should not exist I can easily revert them, so please let me know :)

pakoito commented on modules/core/arrow-test/src/main/kotlin/arrow/test/laws/HashLaws.kt:
```diff
@@ -0,0 +1,30 @@
+import io.kotlintest.properties.forAll
+
+object HashLaws {
+  fun <F> laws(HF: Hash<F>, EQ: Eq<F>, cf: (Int) -> F): List<Law> =
```
Nit: empty line here.

pakoito commented on modules/docs/arrow-docs/docs/docs/typeclasses/hash/README.md:
````diff
@@ -51,6 +51,11 @@ val user = User("MyId", "MyName")
+```
+
+```kotlin:ank
+data class User(val id: String, val name: String)
````
Same as with the imports, there's no need to redefine the class either!

raulraja replied:
The ones that are playable which if possible should be most of the ones used to document API require all the context in which they run in them

1Jajen1 replied:
Ok I'll change that then. But that only goes for context as in classes and variables, not imports right?

raulraja replied:
See the `arrow-query-language` docs for examples. You have to include everything in the snippet that it needs to run. All imports, a main method and a couple of comments that delimit the code you are not collapsing in the snippet when presented on the site.

1Jajen1 replied:
Done (I hope) :)

pakoito commented:
If you'd like to run detekt to find errors like the one in CI, you have to do it on a clean build:

`./gradlew clean detekt`

```
UnusedImports - [option.kt] at modules/core/arrow-instances-core/src/main/kotlin/arrow/instances/option.kt:7:1
```

1Jajen1 commented:
Yeah, will do from now on. Takes a while to get used to a different workflow

raulraja commented on modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Hash.kt:
```diff
@@ -0,0 +1,45 @@
+package arrow.typeclasses
+
+/**
+ * A type class used to represent hashing for objects of type [F]
```
We should probably explain what a Hash is here

raulraja replied:
What a hash value is and what they are useful for. A simple introductory paragraph should be enough IMO.

1Jajen1 replied:
Added something. Not sure if enough, or good. :see_no_evil:

raulraja commented on modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Hash.kt:
```diff
@@ -0,0 +1,45 @@
+   * @receiver The object to hash
+   * @returns an int representing the object hash
+   */
+  fun F.hash(): Int
```
Use kdoc embedded ank snippets to show each one of the public api methods with examples. In the next version of Arrow these will show up in the browser as runnable:
```
/**
   * Commented method or class
   * 
   * {: data-executable='true'}
   *
   * ```kotlin:ank
   * import arrow.aql.instances.list.select.*
   * import arrow.aql.instances.listk.select.select
   *
   * fun main(args: Array<String>) {
   *   //sampleStart
   *   val result: List<Int> =
   *     listOf(1, 2, 3).query {
   *       select { this + 1 }
   *     }.value()
   *  //sampleEnd
   *  println(result)
   *  }
   *  ```
   */
```

raulraja replied:
I ment actually in the sources too. The example will then show up in the API docs beside the docs page.

1Jajen1 replied:
Yeah, that confused me a bit :) 
Added it now for `F.hash()`, `Hash.invoke()` and `Hash.any()`

raulraja commented on modules/docs/arrow-docs/docs/docs/typeclasses/hash/README.md:
````diff
@@ -0,0 +1,95 @@
+```kotlin:ank
+import arrow.core.*
+
+// This will fail because it's not evaluated for hashing
````
```suggestion
// This will return `false` because it's not evaluated for hashing
```
Because if this really fails blowing up the stack we have something wrong in our code.

pakoito commented:
@1Jajen1 is this good to go?

1Jajen1 commented:
If the docs are good then yes. They compile and look alright, but I feel like with the runnable docs you never know (or is there a good way to test them locally with new features)

pakoito commented:
Good for me :D Thanks for your first contribution, I hope it was interesting! If you'd like to take other tickets just let us know.

1Jajen1 commented:
It definitly was. It was something new, I learned a lot and it'll definitly change how I set up my projects in the future. Lots of new tools that feel so much better than what I usually use at work (which is no wonder considering how old codebases tend to work :/). I'll definitly continue, work and university are not that busy atm. The more to learn the better :)

raulraja commented:
There seems to be something wrong with this function. After it has been introduced it always bails with the same error https://api.travis-ci.org/v3/job/457554572/log.txt

1Jajen1 commented:
An exact copy of this runs fine locally (without ank). Thats the example from the source right? Have you ever encountered errors like this before? To me it seems that the way ank runs this somehow breaks it. I can't imagine that snippet breaking memory limits, as its literally just creating an object of Hash<String> and running its extension function. The created object is almost the same as the String.hash() instance (except for the eq part) so I'd suspect ank, but then again I have no idea how it works internally so I might be wrong :confused:

1Jajen1 commented:
So correct me if I am wrong: The way ank works top level is =
- create target dir
- get list of all relevant files
- read all those files
- preprocess all files
- magic
- output

Is there any reason to do those sequential, having all files and their content loaded in memory, as opposed to doing this whole workflow for each file one by one?

pakoito commented:
It may be because it's importing all instances, instead of just the necessary ones? There's a lot of code there.

1Jajen1 commented:
While it might make the problem worse I don't think its the cause. There is a lot of code in the docs that uses wildcard imports, that has previously worked fine. (The docs for equal use exactly the same import)

raulraja commented:
@1Jajen1 It might be very well a problem with Ank. I wasn't built to process that many api docs files. I'm pretty sure that is the issue. Forget about this and I'll try to fix it in that branch before pushing to master
